Block reward contract

andresilva opened the pull request:
Add support for having a contract calculate the block rewards. The engine passes a list of benefactors and reward types to this contract which then returns a list of addresses and respective rewards. This is the contract solidity interface:

```solidity
interface BlockReward {
    // produce rewards for the given benefactors, with corresponding reward codes.
    // only callable by `SYSTEM_ADDRESS`
    function reward(address[] benefactors, uint16[] kind)
		external
		returns (address[], uint256[]);
}
```

Currently I only added support for this in the `AuthorityRound` engine.
Closes #7879.

rphmeier commented on ethcore/res/authority_round_block_reward_contract.json:
```diff
@@ -0,0 +1,61 @@
+	"9cce34f7ab185c7aba1b7c8140d620b4bda941d6": { "balance": "1606938044258990275541962092341162602522202993782792835301376", "nonce": "1048576" },
+	"0000000000000000000000000000000000000042": {
+	  "balance": "1",
+	  "constructor": "6060604052341561000f57600080fd5b6102b88061001e6000396000f300606060405260043610610041576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff168063f91c289814610046575b600080fd5b341561005157600080fd5b610086600480803590602001908201803590602001919091929080359060200190820180359060200191909192905050610125565b60405180806020018060200183810383528581815181526020 ...
```
I guess it's not possible to add a comment here linking to some source? IMO we should keep test specs as localized as possible instead of putting them all in the `res` folder. Then we can annotate the loading code with a link to the contract source.

rphmeier commented on ethcore/src/engines/block_reward.rs:
```diff
@@ -0,0 +1,170 @@
+use error::Error;
+use machine::EthereumMachine;
+
+pub type SystemCall<'a> = FnMut(Address, Vec<u8>) -> Result<Vec<u8>, String> + 'a;
```
grumble: missing docs on this and other members of the module

rphmeier commented on ethcore/src/engines/block_reward.rs:
```diff
@@ -0,0 +1,170 @@
+			.map_err(Into::into)
+			.map_err(::engines::EngineError::FailedSystemCall)?;
+
+		// since this is a non-constant call we can't use ethabi's function output
```
is there an issue filed about that?

andresilva replied:
I don't think so, there's also no general way to do it since it involves creating a transaction and the return data is not included in the receipt (we could trace the transaction).

andresilva replied:
The reward contract might have internal state that needs to be updated.

rphmeier replied:
which can be used to track total supply if you have a reward contract from the beginning. Or even just to maintain a record of recent rewards if you fork to include one.

rphmeier commented:
We should absolutely link to the source of the contract used as close as possible to the tests themselves.

minor grumble: a transition or `activate_at` parameter for the reward contract.

andresilva commented:
I've added the missing docs and added a transition block parameter to enable the reward contract. The example block reward contract that's used in the tests is here: https://github.com/parity-contracts/block-reward/blob/master/contracts/ExampleBlockReward.sol (I added the link in a comment where the spec is loaded). Currently the test spec is shared for the block reward contract client and also for an aura test. 

I agree that it would be better to not have all the test specs in a single folder, maybe we can split this up: https://github.com/paritytech/parity/blob/master/ethcore/src/spec/spec.rs#L809 (also it seems like they're being included in the binary 😟). Should we deal with that in another PR?

rphmeier commented:
yeah, that's beyond the scope of this one but definitely worth a cleanup.

andresilva commented on ethcore/src/engines/block_reward.rs:
```diff
@@ -0,0 +1,184 @@
+/// different semantics which could lead e.g. to different reward values.
+#[repr(u8)]
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum RewardKind {
```
We already have a `RewardType` which is used for tracing but it only distinguishes between `Block` and `Uncle` rewards.

andresilva commented:
Tracked in #8424.

andresilva commented:
Yes it's not possible for the contract to emit events, this is a limitation of how we currently deal with "system" transactions (hopefully this will be fixed in the future).

5chdn commented:
Needs docs :p
